### ISSUE TITLE
Modularize SAM Template file

### ISSUE THREAD

**Issue by Dianadec** [bcgov/reserve-rec-api#37]
Our `template.yaml` file contains all the stack-building instructions for our AWS SAM deployments. Because it is a single file, it can get quite large as it scales with the product size. It is good-boy code protocol to look for ways to break down large file sizes and store the information in them into their own respective modules.

AWS SAM provides a way to do this in the form of [application nesting](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-template-nested-applications.html), however, it has limited support for our application structure. Namely, it cannot have references to the same `RestApi` resource in multiple files, which means our Lambda instructions cannot be separated in different files. Lambda instructions make up the majority of the `template`, so the inability to separate these instructions makes application nesting pretty useless.

There are two workarounds we could investigate:
 - Create a script that builds a singular `template.yaml` file from a bunch of modules at runtime. Before running `sam build`, the script would have to be run to prebuild the instruction file.
 - Investigate AWS CDK, which provides a CI/CD framework with a much higher fidelity than AWS SAM, but takes more time to set learn/set up/maintain, and would take some time to translate our existing SAM work. Note - this is a lot of work but we should be trending to use CDK > SAM anyways.

The longer we wait to do this, the bigger our `template.yaml` gets, and the more work it is to change the file structure when it gets to be too much. 

#### Acceptance Criteria:
 - investigate the pros/cons of each method
 - implement if feasible/appropriate at this time

#### Development Checklist:

- [x] investigate pros/cons of each method
- [ ] if building manual script, build the script and document. 
- [x] if investigating AWS CDK, provide thorough reasoning to pursue this method, and build proof of concept. create another ticket to do the full implementation
- [ ] Make sure all aspects of CI/CD pipeline work (tests, GH actions, package.json scripts, etc) following any changes

#### Dependencies

- Blocked by
- Blocking

**Relevant documentation as reference**

**Definition of Ready**

- [ ] Acceptance criteria are included
- [ ] Wireframes are included (if applicable)
- [ ] Design / Solution is accepted by Product Owner (if applicable)
- [ ] Dependencies are identified (technical, business, regulatory/policy)
- [ ] Story has been estimated (under 13 pts)

**Definition of Done** 
- In progress:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] UI meets accessibility requirements
  - [ ] Unit tests are written
  - [ ] Work is traceable in GitHub
  - [ ] PR linked to ticket number
  - [ ] If needed/required - Dev adds flag/label to highlight any migration steps necessary prior to PROD deployment
- Code review:
  - [ ] Code is peer reviewed and has passed CI/CD tests
- QA:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Code is potentially shippable to the production environment
  - [ ] Functional features have been tested and passed by QA
  - [ ] UI components tested by designer
  - [ ] Code is deployed to PROD when moved to 'done' column (unless requested otherwise by PO)
- PO Review:
  - [ ] Acceptance criteria are tested (Functionality meets the acceptance criteria defined in the ticket)
  - [ ] Reviewed and approved by Product Owner

#### Notes:
- 
-

**Comment by cameronpettit** [comment]
After investigating this for a few days, I have arrived at the following conclusion:

Manually creating a script to compile a SAM `template.yaml` file from a handful of template shards is very similar in process to migrating to AWS CDK - in fact, it is so similar that NOT using AWS CDK in this scenario would be an anti-pattern.

AWS CDK (Cloud Development Kit) is a CICD framework for deploying AWS CloudFormation infrastructure using familiar programming languages. AWS SAM (what this team currently uses) is also a CICD framework offered by AWS - in fact, AWS SAM is just a heavily simplified version of AWS CDK. Both SAM and CDK produce CloudFormation Templates when run, which are in-turn used to deploy the CloudFormation infrastructure. The main difference between the two is just the simplicity of SAM over CDK, while CDK offers greater control over infrastructure rules.

The major drawback of SAM is its lack of modularity. All interrelated CloudFormation infrastructure must exist in a single SAM template, meaning template size can balloon quickly in larger projects. Our current SAM deployments in A&R, DUP, and the Data Register are relatively small, so the size of `template.yaml` has not been an issue in maintainability yet. With `reserve-rec` (PRDT), the `template.yaml` is already well over 1k lines, which poses a maintainability hurdle as the project grows. To be clear - the whole project could still be contained in a single `template.yaml` file and function properly, but as it grows, it becomes exponentially difficult for developers to edit, and slows down local testing due to long `sam build` times. I believe SAM is a powerful tool for quick deployment of small/moderately sized services, but it will struggle to support a project the size of the projected PRDT scope. 

AWS CDK offers far greater flexibility in file management and deployment control at the expense of a more complicated framework. The total LOC of a CDK deployment is far greater than its SAM equivalent, but has benefit of being spread amongst any number of files. In addition, the CDK framework can be written in several modern languages, including `javascript/typescript` which is more familiar to our devs than `YAML`. 

A few more nice things about CDK - local development patterns barely change for developers. Developers can build a CDK deployment locally using `cdk synth`, analogous to `sam build`, and can run APIs locally using `sam local start-api` same as SAM projects. Deploying to a remote AWS env is simple with `cdk deploy`, analogous to `sam deploy`. Project directory structure remains largely the same between SAM and CDK, so the only code changes that need to happen are CI/CD related - Lambda functions, layers, etc remain the same. GitHub Action pipelines would need only minor changes to deploy with CDK. 

CDK templates can be easily generated from existing SAM templates OR CloudFormation Stacks: https://docs.aws.amazon.com/cdk/v2/guide/migrate.html#:~:text=Migrate%20from%20an%20AWS%20SAM,migrate%20from%20the%20deployed%20stack.

1.  Install AWS CDK package `npm install -g aws-cdk`
2. Log into Nimbus, paste credentials into terminal
3. `cdk migrate --from-stack --stack-name "reserve-rec"`

I've done enough exploring with AWS CDK the past few days to be convinced that a migration is easily possible, albeit a little time consuming. Additionally, we would have to spin down the existing SAM deployment simultaneously, and since that is currently supporting some front end work, I haven't gone through with this change yet. 

Currently our `reserve-rec` deployment using SAM is serving us fine but there will come a point where our rate of growth will be impeded by the monolithic structure of SAM templates - and by that time, the effort required to migrate to CDK will have grown exponentially. I recommend we set aside some time soon to do this migration, as it will only get more difficult the more we progress. I consider this issue tech debt - we *should* do this, it's just a matter of when. In the upcoming weeks there is a no-meeting week and the Christmas holidays which would perhaps be opportune times for completing this work.

**Comment by Dianadec** [comment]
Thanks for capturing these thorough notes, @cameronpettit ! So from my understanding, the sooner we we migrate from SAM to CDK, the better. Due to the size of this product, SAM won't be able to handle it well down the road, and we AWS CDK will be preferred from a technical standpoint. I like your idea of the timing for this, being this month. I'll tag @LindsayMacfarlane here as well for her input.

**Comment by Dianadec** [comment]
@cameronpettit shall I make a new ticket to capture the actual work of the AWS CDK migration? And move this ticket along as an exploration ticket?

**Comment by cameronpettit** [comment]
@Dianadec yep, I'd say this ticked is done from an exploration standpoint. If we actually want to go ahead with it we should make a ticket and bring it up in refinement for others to consider the ramifications.

**Comment by Dianadec** [comment]
Thanks Cam! I'll make a ticket for the next step. Moving to PO review.

**Comment by LindsayMacfarlane** [comment]
Approved. 

And, as discussed in refinement, I support your recommendation @cameronpettit to prioritize the switch from SAM to CDK for reservations next sprint.

cc @Dianadec

**Comment by cameronpettit** [comment]
@LindsayMacfarlane a little more context in case you need any:

The first time AWS infrastructure was used by this team was for DUP back in 2021. Our infrastructure was deployed by a Terraform script that was written by our then-devops developer in that really small 5 week window. The script worked, but was hardly scalable. A lot of effort had to be put in to make any changes to our infrastructure. 

Restrictions imposed by the AWS team prevented us from moving away from Terraform by the time our next project, A&R came around. After A&R was launched there were changes made that allowed us to move away from Terraform, which was hampering development greatly by that point, so we moved to the quickest alternative, which was SAM. We'd spent a bit of time investigating the feasibility of SAM because we were so fed up with Terraform so we jumped at the opportunity to use it with the Data Register. SAM's been working for us great ever since. So great that we backported it to DUP and A&R so we could make changes to infrastructure quickly on those projects.

So that's a brief history of why we are on SAM for our projects right now and didnt jump directly to CDK.
